### ISSUE TITLE
"Authentication" Link in Admin Portal Fails with 400 Bad Request After Migrating to Version 25

### ISSUE THREAD

**Issue by bclements** [keycloak/keycloak#32624]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/ui

### Describe the bug

**Summary:** 
After performing a stepwise migration of a Keycloak server from version 21 to the latest version (25), the "Authentication" link in the admin portal causes a 400 Bad Request error. 

**Reproduction Rate:**

100% on upgraded instances (from version 21 to 25) with existing database data.
0% on fresh Keycloak instances that follow the same upgrade path without existing data.

**Environment:**

- Keycloak Version: 25, also tried nightly
- Previous Versions Tested: 21, 22, 23, 24 (no issues with the "Authentication" link)
- Deployment Environment:
- Production: AWS with ALB, Local Development: localhost:8080 (using start command)
- Configuration:
-- hostname-strict=false
-- proxy-headers=xforwarded and also tried proxy-headers-forwarded

![Error Message 1](https://github.com/user-attachments/assets/9c8b5e3e-7d55-4e3c-8f20-35b9085427c8)

**Investigation Findings:**
Please note that another [github issue ](https://github.com/keycloak/keycloak/issues/32531) noted a similar error and the solution was to make sure the "custom providers" were installed. Our custom providers have followed each upgrade from 21->22->23->24 and every time, the authentication link works. It's when we move to 25 or latest, this issue occurs. 

The issue is only reproducible in instances that have been upgraded from version 21 to 25 with existing database data.
Viewing the admin ui using chrome developer tools, I see the the "required-actions" endpoint returns a 400 - bad request almost immediately on clicking the "Authentication" link.
![Required Actions Resource](https://github.com/user-attachments/assets/3db80a62-0066-41ff-9ea5-ad6a19170347)
![Required Actions Response error](https://github.com/user-attachments/assets/1ba2945b-9441-441c-b634-c03349ba8852)

Here is the relevant snippet from the debug log:
`2024-09-03 17:23:30,078 DEBUG [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-14) Error response 400: com.fasterxml.jackson.databind.JsonMappingException: Cannot invoke "org.keycloak.authentication.RequiredActionFactory.isConfigurable()" because "factory" is null
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._wrapAsIOE(DefaultSerializerProvider.java:531)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:504)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:341)
        at com.fasterxml.jackson.databind.ObjectWriter$Prefetch.serialize(ObjectWriter.java:1574)
        at com.fasterxml.jackson.databind.ObjectWriter._writeValueAndClose(ObjectWriter.java:1275)
        at com.fasterxml.jackson.databind.ObjectWriter.writeValue(ObjectWriter.java:1098)
        at io.quarkus.resteasy.reactive.jackson.runtime.serialisers.FullyFeaturedServerJacksonMessageBodyWriter.writeResponse(FullyFeaturedServerJacksonMessageBodyWriter.java:79)
        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:217)
        at org.jboss.resteasy.reactive.server.core.ServerSerialisers.invokeWriter(ServerSerialisers.java:185)
        at org.jboss.resteasy.reactive.server.core.serialization.FixedEntityWriter.write(FixedEntityWriter.java:28)
        at org.jboss.resteasy.reactive.server.handlers.ResponseWriterHandler.handle(ResponseWriterHandler.java:34)
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:147)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:635)
        at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.NullPointerException: Cannot invoke "org.keycloak.authentication.RequiredActionFactory.isConfigurable()" because "factory" is null
        at org.keycloak.admin.ui.rest.AuthenticationManagementResource.toRepresentation(AuthenticationManagementResource.java:144)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:601)
        at com.fasterxml.jackson.datatype.jdk8.StreamSerializer.serialize(StreamSerializer.java:71)
        at com.fasterxml.jackson.datatype.jdk8.StreamSerializer.serialize(StreamSerializer.java:15)
        at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:502)
`
**Suggested Priority:** High - This issue affects the ability to manage authentication settings in the admin portal, which is a crucial functionality for maintaining security and access control.

**Acceptance Criteria:**

The "Authentication" link in the admin portal should work as expected after migrating from version 21 to version 25 with existing database data.
Ensure that the functionality is consistent across both upgraded and fresh instances.

### Version

nightly

### Regression

- [X] The issue is a regression

### Expected behavior

**Expected Behavior:** The "Authentication" link should function correctly and load the authentication management page without any errors, consistent with previous versions up to 24.

### Actual behavior

**Actual Behavior:** Clicking on the "Authentication" link results in a 400 Bad Request error as indicated in the browser's developer tools. This issue does not occur when using a fresh Keycloak instance deployed from version 21 to 25 following the same upgrade path. 

### How to Reproduce?

**Steps to Reproduce:**
Please Note: This issue is observed only after the migration to version 25 using our database/existnig data and does not occur in a fresh deployment following the same upgrade path. The problem appears to be linked to the upgraded database configuration.
1. Start with an upgraded Keycloak instance, initially on version 21, with existing database data.
2. Migrate the Keycloak server stepwise through major versions (21 -> 22 -> 23 -> 24 -> 25).
3. Configure Keycloak with the following settings:
- hostname-strict=false
- tried both proxy-headers=xforwarded and proxy-headers=forwarded in different test cases

4. Access the admin portal and navigate to various menu items within any realm to verify general functionality.
5. Click on the "Authentication" link in the admin portal.

**Comment by bclements** [comment]
@thomasdarimont I believe the code you added 5 months ago to make required actions configurable is surfacing this error. 

I'm adding some debugging to your code in various places to try to figure out what in my database or what provider is causing the issue. I'll post my findings once I get back results.

**Comment by thomasdarimont** [comment]
I think the issue is that your realm configuration contains a required action provider reference that no longer exists. Either the provider implementation was removed or the alias changed.

I think we can just add a null check there and skip "broken" required action references. We should also print a warning if an unknown required action reference is found.

**Comment by thomasdarimont** [comment]
I added a null guard to prevent NPE in #32649

**Comment by bclements** [comment]
We tested this fix, along with the debugging additions in a local dev account and it works great. Was able to find out that a previous preview feature that we disabled back in 21 was following us throughout our upgrade path.

**Comment by ssilvert** [comment]
@thomasdarimont @bclements I removed the `backport/24.0` label because the backport for Keycloak 24 looked quite complicated with conflicts.

The backport to Keycloak 24 branch was straightforward.  If you guys think it is really important to backport to 24 we can take a second look.

**Comment by bclements** [comment]
backport to just 25 is perfectly fine.

**Comment by ben-kiplot** [comment]
In case it helps anyone else with the same issue, this issue was caused for us by adding `web-authn` to `KC_FEATURES_DISABLED` after it was previously enabled, resulting in web authn-related required action providers remaining in the configuration.

**Comment by parvez-probuy** [comment]
Thanks everyone. I looked through the issues and chats and followed the steps to find the root cause for my issue.
In my case, it was the "idp_link" required actions.

For anyone wondering, I used Keycloak REST APIs and followed these steps 
1. Checked the items in "required-action" from /admin/serverinfo API
2. Fetched required-actions for each realm using /admin/realms/:realm/authentication/required-actions API
3. Cross verified both lists and checked for any missing items from 1st list (serverinfo "required-actions") which are present in 2nd list (realm/required-actions API).
4. Delete the extra items which are in 2nd list and not in 1st list. (using /admin/realms/:realm/authentication/required-actions/:alias API)

It should work! Kudos!
